[#146] update user details endpoint plus tests backup

ratulkai93 opened the pull request:
This endpoint will allow users to update/modify their password and organization name inside of the database.

ratulkai93 committed:
[146] added interface for user detail updater

ratulkai93 committed:
[146] added user detail model for updating password/organization

ratulkai93 committed:
[146] added user details validation logic in authentication service

validateUserDetails method checks for existing email and proceeds to change organization name and password

ratulkai93 committed:
[146] added methods for sending request to update user details on the controller layer

updateUserDetailsRequest can be accessed through the router and called
if the request is valid, this method will allow user to change/update password and organization name through service layer logic

ratulkai93 committed:
[146] added working router path

its now possible to successfully update password/organization name (can be tested using postman)

ratulkai93 committed:
[146] added tests for updateUserPassword/OrganizationName from authentication model

ratulkai93 committed:
[146] added test for valid user detail update request

ratulkai93 committed:
minor adjustments

ratulkai93 committed:
Merge remote-tracking branch 'origin/develop' into 146-Update-user-details-endpoint-plus-tests-backup

ratulkai93 committed:
[146] added minor fixes for tests

ratulkai93 committed:
[146] added additional controller tests for update user details

GD-Dheer commented on Server/src/genericInterfaces/AuthenticationInterfaces.ts:
```diff
@@ -18,6 +18,12 @@ export interface IJwtParams {
     firstName: string
 }
 
+export interface IUserDetailUpdater {
```
Its confusing with the other interface below, maybe we can call this IUpdateUserDetail, and the one below IFetchUserDetail

ratulkai93 replied:
fixed in the latest commit

GD-Dheer commented on Server/src/models/AuthenticationModel.ts:
```diff
@@ -127,4 +127,33 @@ export class AuthenticationModel {
     }
+
+
+    //for issue 146
```
Do we need this comment?

ratulkai93 replied:
removed all unnecessary comments on the recent commit

GD-Dheer commented on Server/src/models/AuthenticationModel.ts:
```diff
@@ -127,4 +127,33 @@ export class AuthenticationModel {
+    static async updateUserPasswordDetail(email: string, passwordHash: string): Promise<boolean> {
+
+        let connection = getConnection();
+        //for password
```
If comments are needed we can add them above the method

ratulkai93 replied:
done

GD-Dheer commented on Server/src/models/AuthenticationModel.ts:
```diff
@@ -127,4 +127,33 @@ export class AuthenticationModel {
+    static async updateUserOrganizationDetail(email: string, organization: string): Promise<boolean> {
+
+        let connection = getConnection();
+        //for organization
```
The comment can be moved above or removed

GD-Dheer commented on Server/src/controllers/authenticationController.ts:
```diff
@@ -64,6 +65,44 @@ export class AuthenticationController {
     }
 
+    private validateUserDetailRequest(request: Request): boolean {
+        if (request.query.hasOwnProperty('password')) {
```
Do we need to check organization here? Or just password?

ratulkai93 replied:
i think just checking for password is fine. Since we already checked for email, if there is no password , it won't make sense to have a valid request. If needed i can add the check for the organization name .

GD-Dheer replied:
What happens if a user is only passing an update to organization without a password update?

ratulkai93 replied:
uhh, if i check off the password field and only input email and organization name, i get an invalid request. however, it works if i leave the password field checked but empty

ratulkai93 replied:
fixed on the latest commit. we can now update the organization name without the password update

GD-Dheer commented on Server/src/models/AuthenticationModel.ts:
```diff
@@ -127,4 +127,33 @@ export class AuthenticationModel {
+
+
+    //for issue 146
+    static async updateUserPasswordDetail(email: string, passwordHash: string): Promise<boolean> {
```
I think this can be called just updateUserPassword, don't think we need detail here and the param can be called newPasswordHash for clarity

GD-Dheer replied:
@ratulkai93 This is still valid

TiraFire replied:
Will this be changed?

ratulkai93 replied:
i believe i fixed it on the recent commit

GD-Dheer commented on Server/src/models/AuthenticationModel.ts:
```diff
@@ -127,4 +127,33 @@ export class AuthenticationModel {
+
+        return true
+    }
+    static async updateUserOrganizationDetail(email: string, organization: string): Promise<boolean> {
```
Same as password comment above

GD-Dheer commented on Server/src/services/authenticationService.ts:
```diff
@@ -119,12 +120,38 @@ export class AuthenticationService {
+        let password = userDetailUpdater.password
+        let organization = userDetailUpdater.organization
+
+        // email = await AuthenticationModel.verifyIfEmailExists(userDetailUpdater.email);
```
Do we need this commented line?

ratulkai93 replied:
removed

GD-Dheer commented on Server/src/services/authenticationService.ts:
```diff
@@ -119,12 +120,38 @@ export class AuthenticationService {
+            }
+        }
+        else {
+            throw new BadRequest("Email is not valid. Please enter the correct email!");
```
We can make this message a bit better since by now the user has already logged in. The only time we say email is not valid - please enter correct is when they are trying to login.

Something like: Email cant be found, request declined - use error 406 maybe

ratulkai93 replied:
i updated the message and the test is now running. Thank you for the feedback ;)

GD-Dheer commented on Server/src/services/authenticationService.ts:
```diff
@@ -119,12 +120,38 @@ export class AuthenticationService {
      * @param userEmail User Email
      */
-    async loadUserDetails(userEmail: string): Promise<IResponse> {
+    async loadUserDetails(userEmail: string): Promise<IResponse> {//need test
```
What is the //nest test?

ratulkai93 replied:
removed

GD-Dheer commented on Server/src/tests/controllers/authenticationController.spec.ts:
```diff
@@ -145,5 +145,51 @@ describe('Authentication Controller', () => {
+    });
+
+    test('invalid user detail update request', async () => {
+        const expectedResponse = "Email is not valid. Please enter the correct email!";
```
This message would need to be updated from above

GD-Dheer commented on Server/src/controllers/authenticationController.ts:
```diff
@@ -64,6 +65,44 @@ export class AuthenticationController {
+        } else {
+            let requestParams: any = { ...request.query };
+            let updateUserDetail: IUserDetailUpdater = requestParams;
+            //call validateUserDetails from service inside callServiceForUpdateUserDetails
```
Comment here?

ratulkai93 replied:
removed

GD-Dheer commented on Server/src/routes/authenticationRouter.ts:
```diff
@@ -25,9 +25,14 @@ router.get('/resetPassword', (request: Request, response: Response, next: NextFu
 
 });
 
+router.post('/updateUserInfo', (request: Request, response: Response, next: NextFunction) => {
```
We need to add a middleware here -  JWTAuthenticator.verifyJWT because this is a protected path. Like the /userDetails

ratulkai93 replied:
updated on the recent commit. thank you for pointing it out

GD-Dheer reviewed (changes requested):
Small stuff to address. Thanks for the work!

GD-Dheer commented:
> Small stuff to address. Thanks for the work!

@ratulkai93 One of your tests is also failing on travis

ratulkai93 committed:
[146] added fixes addressing PR comments

-fixed the controller test for user details missing password/organization Name
-added middleware to the the router
-renamed interfaces for ease of access
-changed message for invalid requests
-removed unnecessary commits

GD-Dheer commented on Server/src/genericInterfaces/AuthenticationInterfaces.ts:
```diff
@@ -18,13 +18,13 @@ export interface IJwtParams {
 }
 
-export interface IUserDetailUpdater {
+export interface IUpdateUserDetail {
```
Nice

GD-Dheer commented on Server/src/models/AuthenticationModel.ts:
```diff
@@ -102,7 +102,7 @@ export class AuthenticationModel {
      * @param email - User Email
      */
-    static async getUserDetails(userEmail: string): Promise<IUserDetails> {
+    static async getUserDetails(userEmail: string): Promise<IFetchUserDetail> {
```
Can you change this one to fetch aswell instead of get?

ratulkai93 replied:
updated on the latest commit

GD-Dheer commented on Server/src/routes/authenticationRouter.ts:
```diff
@@ -25,10 +25,11 @@ router.get('/resetPassword', (request: Request, response: Response, next: NextFu
 
-router.post('/updateUserInfo', (request: Request, response: Response, next: NextFunction) => {
-    //make an updatePasswordRequest and validatePasswordRequest 
+router.post('/updateUserInfo', JWTAuthenticator.verifyJWT, async (request: Request, response: Response, next: NextFunction) => {
```
Could you test this again. To make sure it is functional.

Steps
1 -> Login with valid user 
2 -> Take the Token response and input in Postman
3 -> Find Authorization Tab, select Bearer Token and input the Token Response
4 -> Call your endpoint again

ratulkai93 replied:
managed to get it working after removing the quotation marks from the token response. thank you!

GD-Dheer reviewed (approved):
Nice updates

ratulkai93 committed:
[146] changed getUserDetails to fetchUserDetails

ratulkai93 committed:
[146] small controller fix

the current update allows user  to update the organization name without requiring to change the password or vice-versa

TiraFire commented on Server/src/controllers/authenticationController.ts:
```diff
@@ -64,6 +65,43 @@ export class AuthenticationController {
+    private validateUserDetailRequest(request: Request): boolean {
+        if (request.query.hasOwnProperty('password') || request.query.hasOwnProperty('organization')) {
+            return true;
+        } else
```
in which cases could this return false?

ratulkai93 replied:
if the request does not have both password and organization name, then it would return false

TiraFire reviewed (approved):
The tests seem to cover everything, I assume the tests work with the test db? Just missing the small name changes requested by @GD-Dheer

ratulkai93 committed:
Merge branch 'develop' into 146-Update-user-details-endpoint-plus-tests-backup

ratulkai93 committed:
[#146] removed comment and merged develop
